### ISSUE TITLE
Multiple search arguments don't work

### ISSUE THREAD

**Issue by Hubro** [jordansissel/xdotool#357]
Command:

```
$ xdotool search --all --class discord --name " - Discord$"
```

Output:

```
xdotool: Unknown command: --name
Run 'xdotool help' if you want a command list
```

This should be valid according to the man page.

**Comment by mdtrooper** [comment]
I have similar situation.

```
$ xdotool --version
xdotool version 3.20160805.1

$ dpkg -l | grep xdotool
ii  xdotool                                                     1:3.20160805.1-4                           amd64        simulate (generate) X11 keyboard/mouse input events

$ uname -a
Linux portatil 6.2.0-26-generic #26~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Jul 13 16:27:29 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by ceejatec** [comment]
FYI your command is actually incorrect according to the manpage, but IMHO the way it works is terrible.

The search command is
```
xdotool search [options] pattern
```
where `[options]` can include two very different kinds of options: normal things like `--onlyvisible` or `--screen 2` which do what you'd expect; **or** `--class`, `--classname`, `--name`, or `--role`, _each of which matches the single specified `pattern`_. Those four options do not accept a pattern of their own as an argument. So it is impossible to do a search that matches the window class "discord" _and_ the window name " - Discord$", because you can only specify one pattern to match. (But see below.)

Again, this is exactly what the manpage specifies; it's just a bizarrely surprising and non-useful thing to do.

Command chaining doesn't work either - you can't do eg.
```
xdotool search --class discord search --name " - Discord$"
```
Or, rather, you _can_ do that, but it will silently drop the first search results and use only the second.

Now, the small saving grace here is that the `pattern` argument _is_ a pattern, and so you can do something a bit absurd like this:
```
xdotool search --all --class --name '(^discord$| - Discord$)' 
```
Which will match both the window class and the window name against the combined regular expression. That will work in many cases, although it feels super-clunky. You'll want to specify your regexes fairly completely, using `^` and `$` where possible, to prevent it accidentally matching some other window that happens to have "discord" in its name, like a browser window open to a webpage about Discord. Bear in mind that the pattern matching is always case-insensitive. Also don't forget to include `--all` so it only selects windows that match both `--class` and `--name`, since the default is to join the options with "or".

(Subsidiary rant: `pattern` is not optional, even if you don't specify any of the four options which make use of it. ie,
```
xdotool search --desktop 0
```
is invalid. You have to specify some pattern, and it will then treat it as though you specified `--class --classname --role --name` and match the pattern against all four pieces of metadata. So if you really do want "all windows on desktop 0", you'll need to do
```
xdotool search --desktop 0 '""
```
or some other pattern that will match every string.)

**Comment by jordansissel** [comment]
I agree the behavior is surprising and I wish I had done it differently. I would fix it, but that would probably break existing search users. I’m open to adding another search command, like ‘search2’ or something… doesn’t feel right but it would get us a clean slate to implement search in a better way?

**Comment by ceejatec** [comment]
Yeah, backwards-compatibility is pretty important here, I imagine. I can envision a couple approaches; I'm not sure if either is better, more useful, easier to implement, etc., but:

### Option 1
Introduce a new command as you said, perhaps `find` or `match`. If you did that, I'd recommend also making `--all` the default, with `--any` available as an option. That would be more generally useful for things like "find one specific window and raise it", which I imagine is a common use case where people stumble on the current `search` implementation. Also, it provides a small additional reason for having a separate command, beyond just "hindsight".

```
xdotool find [option [option ...]]

The available options are:
    --class <pattern>
    --classname <pattern>
    --role <patern>
    --name <pattern>
    ...
```
where the remaining options would be identical to `search`, except `--all` would be the default, and you could drop the deprecated `--title`.

### Option 2
Allow command-chaining to work for multiple searches, where each subsequent `search` would only match on the current window stack. That would _technically_ be backwards-incompatible, but only in the sense that something that served no purpose in the past would now serve a purpose, so I can't imagine it would trip too many people up. So eg. this would work:

```
xdotool search --class discord search --name " - Discord$"
```
and would only list window IDs that matched both criteria.

### ---

I personally prefer Option 1, since Option 2 by itself kind of just adds to the confusion ("Why does `search --desktop 0 --class discord` only need one "search" but `search --class discord search --name " - Discord$"` need two?"). But both could resolve this particular complaint at least.

And even with Option 1, it might be kind of neat if command-chaining worked with both `search` and `find`; you could do some stunts like

```
xdotool search --class xterm windowsize 80 24 search --name Primary windowfocus
```
to resize all xterms and then focus the "Primary" xterm in a single command. Maybe more "clever" than "useful", but hey.

**Comment by ceejatec** [comment]
Random aside: `man xdotool` here has this example under "COMMAND CHAINING":
```
xdotool search --class xterm -- windowsize --usehints %@ 80 24
```
which implies that `--` by itself can be used to separate the steps of a command chain. That doesn't actually work, but it'd be kind of nice if it did! Especially if command chaining worked on `search` or `find`, it make things like
```
xdotool search --class discord -- search --name " - Discord$" 
```
a little easier to parse out.

**Comment by t-oster** [comment]
Hi,
I just stumbeled on this exact issue. I think it is a very common use case to search for a window by name AND classname. So I would vote for the Option 1, introduce a new find-command with the expected syntax and --all being the default. Then maybe deprecate the search command in favor of the find command.

**Comment by GoodbyeNJN** [comment]
Has this issue advanced? I am very eager for this feature.

**Comment by jordansissel** [comment]
I’m in favor of option 1 above, a new ‘find’ command that deprecates ‘search’ and operates as described above. 

I haven’t done any work on this, yet.

**Comment by osirisgothra** [comment]
☑️  yes, I defintely agree on option 1. That is because its current state, as "weird" as it may be, is still needed for backwards compatibility. Many including myself have scripts that expect the current behavior, and that's just a can of worms we don't need to open up 🙀
